### PR TITLE
[codegen] Initial version of Cluster code generation.

### PR DESCRIPTION
**Pull request by turon**
Initial version of using Matter SDK code generation to create the matter.js cluster definitions.
This doesn't generate perfect output yet, but provides an initial framework for automatic generation of the cluster handlers.

### CODE REVIEW

**Comment by Apollon77** [comment]
@turon Thats cool. It would - in any case - simplify the initial creation of the files which is basically 60-70% of the work. The rest is then specs reading and potential adjustments ... or do we trust the XML 1000%? I found several things in the XMLs that do not make sense sometimes ... if they are the truth then  the specs are wrong in several places :-)

Maybe let the two of us have a call next week to discuss the goal and next steps. In anycase thtis is cool. I will review the generation template later today!

File: tools/codegen/matter_idl_plugin/__init__.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  if zap_type_lower in float_types:
+    return "TlvFloat"
+  if zap_type_lower == "double":
+    return "TlvDouble"
+  if zap_type_lower == "boolean":
+    return "TlvBoolean"
+  if zap_type_lower in string_types:
+    return "TlvString"
+  if zap_type_lower in bytes_types:
+    return "TlvBbytes"
```
Comment (Apollon77) [inline tools/codegen/matter_idl_plugin/__init__.py:113]: Should be TlvByteString I think

**Comment by turon** [comment]
There is a bug in the SDK that caches an older version of the matter_idl
lib in the pigweed .environment.  i should add a note to the README to
delete that to make it work.

File: tools/codegen/matter_idl_plugin/matter_js_schema.jinja [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+export const enum {{entry.name}} {
+    {%-  for field in entry.entries %}
+    {{field.name}} = {{field.code}},{%   endfor %}
+}
+{%   endfor %}{% endif -%}
+{%- if cluster.bitmaps %}
+
+  // Bitmaps
+{%-   for entry in cluster.bitmaps %}
+export const enum {{entry.name}} {
```
Comment (Apollon77) [inline tools/codegen/matter_idl_plugin/matter_js_schema.jinja:29]: BitMaps should look like

```
const TlvNameSupportBitmap = TlvBitmap(TlvUInt8, {
    groupNames: BitFlag(7),
});
```

File: tools/codegen/matter_idl_plugin/matter_js_schema.jinja [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import { TlvEnum, TlvUInt16, TlvUInt64 } from "../tlv/TlvNumber.js";
+import { TlvString } from "../tlv/TlvString.js";
+import { TypeFromSchema } from "../tlv/TlvSchema.js";
+import { TlvBoolean } from "../tlv/TlvBoolean.js";
+
+{%- if cluster.enums %}
+
+  // Enums
+{%-   for entry in cluster.enums %}
+export const enum {{entry.name}} {
```
Comment (Apollon77) [inline tools/codegen/matter_idl_plugin/matter_js_schema.jinja:20]: should we export by default?

File: tools/codegen/matter_idl_plugin/matter_js_schema.jinja [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    {{field.name}} = {{field.code}},{%   endfor %}
+}
+{%   endfor %}{% endif -%}
+{%- if cluster.structs %}
+
+  // Structs
+{%-   for entry in cluster.structs %}{% if not entry.tag %}
+const Tlv{{entry.name}} = TlvObject({
+    {%-  for field in entry.fields %}
+    {{field.name}}:  TlvField({{field.code}}, {{field | toTlvSchemaFullType}}),{%   endfor %}
```
Comment (Apollon77) [inline tools/codegen/matter_idl_plugin/matter_js_schema.jinja:40]: could we also get min/max and the others in here? Or if not programmatically maybe as comment?

File: tools/codegen/matter_idl_plugin/matter_js_schema.jinja [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ */
+export const {{cluster.name}}Cluster = Cluster({
+    id: {{cluster.code}},
+    name: "{{cluster.name}}",
+    revision: 1,
+{%- if cluster.attributes %}
+
+    attributes: {
+{%-   for attr in cluster.attributes %}
+        {{attr.definition.name}}: Attribute({{attr.definition.code}}),  {{attr.definition.data_type.name | toTlvSchemaType}} }),
```
Comment (Apollon77) [inline tools/codegen/matter_idl_plugin/matter_js_schema.jinja:69]: Same here for optional and writable ...

File: tools/codegen/matter_idl_plugin/matter_js_schema.jinja [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+{% endfor %}
+  }{% endif %}
+{%- if cluster.commands %}
+
+    commands: {
+{%-   for cmd in cluster.commands %}
+        {{cmd.name | toLowerCamelCase}}: Command({{cmd.code}}, Tlv{{cmd.name}}Request, {{cmd.code}}, Tlv{{cmd.name}}Response),
+{% endfor %}
```
Comment (Apollon77) [inline tools/codegen/matter_idl_plugin/matter_js_schema.jinja:77]: could you also add the events?

**Review by Apollon77 — COMMENTED** [review]
I will just leave my thoughts as comments because question is about the goal. this is a cool basis, but also a lot is not yet implemented which needs manual work afterwards in any case ... but would save time already

**Comment by turon** [comment]
> or do we trust the XML 1000%? I found several things in the XMLs that do not make sense sometimes ... if they are the truth then the specs are wrong in several places :-)

The XML is somewhat the truth-on-the-ground because the Matter SDK is what is used to build and certify final products and it uses the XML codegen extensively for both client and server code.  Any spec bugs should be captured as an issue and we can correct it (most likely by correcting the spec to match the 1.0 SDK behavior at this point).  The XML lacks certain things where the manually crafted definitions will remain better such as `@see` spec references and documentation.

**Comment by Apollon77** [comment]
Ok, then I can go through the XMLs and specs of the clusters we have and will open issues for all things I see differences ...

**Comment by Apollon77** [comment]
based on latest learnings this is true, BUT the following things are nmot covereed by the XMLs:
* default values have a different meaning as in specs
* Non-volatile flag is not in XML

File: tools/codegen/matter_idl_plugin/matter_js_schema.jinja (hunk 3 above)
Comment (turon) [reply tools/codegen/matter_idl_plugin/matter_js_schema.jinja:21]: Looks like enums in hand-written clusters are all exported, so yes.

**Comment by Apollon77** [comment]
@turon I will close here in favor of #192 . Thank you for your collaboration!

**Pull request closed without merging**